automata/fa/dfa.py

```diff
@@ -467,17 +467,44 @@ def complement(self):
         new_dfa.final_states ^= self.states
         return new_dfa
 
+    def _get_reachable_states_product_graph(self, other):
+        """Get reachable states corresponding to product graph between self and other"""
+        assert self.input_symbols == other.input_symbols
```

Comment:
@eliotwrobson I'd prefer to raise some new or existing subclass of `AutomatonException` to handle this check, rather than an `assert`. Can you please add a new `SymbolMismatchError` exception class in `automata/base/exceptions.py` and use that instead? The docstring can be something like `"""The input symbols between the given automata do not match"""`.

To clarify, I'm thinking code like the following:

```python
class SymbolMismatchError(AutomatonException):
    """The input symbols between the given automata do not match"""
    pass
```

```python
if self.input_symbols != other.input_symbols:
  raise SymbolMismatchError('The input symbols between the two given DFAs do not match')
```

Please also update the README to document this new `SymbolMismatchError` exception class.

Thanks,
Caleb

Reply:
Sounds good! There are a couple of places actually that a similar assert gets used, so I'll go ahead and replace them all.

Reply:
@eliotwrobson Great! Just comment here when you're done with your remaining changes to this PR. It is looking great so far.

Reply:
@caleb531 Just took another look and should be good to go. I realize I keep making incremental changes to the same functions in different PRs. Prepping these changes to migrate over to this repo gave me a chance to see how everything could be optimized even more, but I think eliminating these data structures was the last real hurdle. A nice side effect has been making our CI job significantly more efficient.

Reply:
@eliotwrobson I like the sound of that! Thank for for these contributions. I will merge now.